code/espurna/sensors/VL53L1XSensor.h

```diff
@@ -0,0 +1,108 @@
+// -----------------------------------------------------------------------------
+// VL53L1X Sensor over I2C
+// Copyright (C) 2017-2018 by Xose Pérez <xose dot perez at gmail dot com>
+// -----------------------------------------------------------------------------
+
+#if SENSOR_SUPPORT && VL53L1X_SUPPORT
+
+#pragma once
+
+#undef I2C_SUPPORT
+#define I2C_SUPPORT 1 // Explicitly request I2C support.
+
+
+#include "Arduino.h"
+#include "I2CSensor.h"
+#include "VL53L1X.h"
+
+class VL53L1XSensor : public I2CSensor {
+
+    public:
+
+        // ---------------------------------------------------------------------
+        // Public
+        // ---------------------------------------------------------------------
+
+        VL53L1XSensor(): I2CSensor() {
+            _count = 1;
+            _sensor_id = SENSOR_VL53L1X_ID;
+            _vl53l1x = new VL53L1X();
+        }
+
+        ~VL53L1XSensor() {
+          delete _vl53l1x;
+        }
+
+        // ---------------------------------------------------------------------
+
+        void setDistanceMode(VL53L1X::DistanceMode mode) {
+          _vl53l1x->setDistanceMode(mode);
+        }
+
+        void setMeasurementTimingBudget(uint32_t budget_us) {
+          _vl53l1x->setMeasurementTimingBudget(budget_us);
+        }
+
+        void setInterMeasurementPeriod(unsigned int period) {
+          if (_inter_measurement_period == period) return;
+          _inter_measurement_period = period;
+          _dirty = true;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        void begin() {
+          if (!_dirty) {
+            return;
+          }
+
+          if (!_vl53l1x->init()) {
+            return;
+          };
+
+          _vl53l1x->startContinuous(_inter_measurement_period);
+
+          _ready = true;
+          _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[20];
+            snprintf(buffer, sizeof(buffer), "VL53L1X @ I2C (0x%02X)", _address);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_DISTANCE;
+            return MAGNITUDE_NONE;
+        }
+
+        // Pre-read hook (usually to populate registers with up-to-date data)
+        void pre() {
+            _error = SENSOR_ERROR_OK;
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index != 0) return 0;
+
+            return _vl53l1x->read() / 1000.0;
```

Comment:
@xoseperez I believe this blocks for a maximum of `_inter_measurement_period` (which should be always as long as the timing budget - I may add this guard). We can also set a timeout. Is there any example for using interrupts for reading data in espurna?

Reply:
hmm. reading about read(), it has an optional arg blocking=true/false, which you do not specify and it defaults to true. _vl531lx->dataReady() is a check for non-blocking mode (which could be in pre() ?) and value() will just return cached value.

Reply:
@mcspr suggestion is better than using interrupts (which you can, check the EventSensor, for instance). Non-blocking calls are always preferred.